### PR TITLE
Nested Blacklist

### PR DESCRIPTION
**Pull request by jason0415**
More test cases:
```
params = {
    data: {
      schema: {
        value: 123,
        value2: 234,
        value3: []
      },
      array: []
    }
  }
```

```
BLACKLIST = ['data.schema.value']
=> {:data=>{:schema=>{:value=>"REDACTED", :value2=>234, :value3=>[]}, :array=>[]}}

BLACKLIST = ['data.schema']
=> {:data=>{:schema=>{:value=>"REDACTED", :value2=>"REDACTED", :value3=>"REDACTED"}, :array=>[]}}

BLACKLIST = ['data.schema', 'data.schema.not_the_key']
=> {:data=>{:schema=>{:value=>"REDACTED", :value2=>"REDACTED", :value3=>"REDACTED"}, :array=>[]}}

BLACKLIST = ['data.schema.foo.bar', 'data.schema.value']
=> {:data=>{:schema=>{:value=>"REDACTED", :value2=>234, :value3=>[]},  :array=>[]}}

BLACKLIST = ['data']
=> {:data=>{:schema=>{:value=>"REDACTED", :value2=>"REDACTED", :value3=>"REDACTED"}, :array=>"REDACTED"}}

```

### CODE REVIEW

**Comment by pedelman** [comment]
Ref. #22

File: lib/gruf/instrumentation/request_logging/hook.rb [hunk 1]
```diff
@@ -139,6 +141,29 @@ def sanitize(params = {})
         def options
           super().fetch(:request_logging, {})
         end
+
+        def redact!(parts, i, params, redacted_string)
```
Comment (pedelman) [inline lib/gruf/instrumentation/request_logging/hook.rb]: Wonder if it makes sense to move this implementation to its own class to simplify testing.

Comment (splittingred) [inline lib/gruf/instrumentation/request_logging/hook.rb]: I think it's probably fine here; adding classes can reduce performance.

Comment (splittingred) [inline lib/gruf/instrumentation/request_logging/hook.rb]: Can you please add docblocks to each method with param/return descriptions and a summary?

File: lib/gruf/instrumentation/request_logging/hook.rb [hunk 2]
```diff
@@ -139,6 +141,29 @@ def sanitize(params = {})
...
+              hash_deep_redact!(params[parts[i]], redacted_string)
+            else
+              params[parts[i]] = redacted_string
+            end
+            return
+          end
+          redact!(parts, i + 1, params[parts[i]], redacted_string)
+        end
+
+        def hash_deep_redact!(hash, redacted_string)
```
Comment (splittingred) [inline lib/gruf/instrumentation/request_logging/hook.rb]: Can you please add docblocks to this method with param/return descriptions and a summary?

File: lib/gruf/instrumentation/request_logging/hook.rb [hunk 3]
```diff
@@ -139,6 +141,45 @@ def sanitize(params = {})
...
+        # Helper method to recursively redact based on the black list
+        #
+        # @param [Array] The blacklist. ex. 'data.schema' -> [:data, :schema]
+        # @param [Integer] The current index of the blacklist
+        # @param [Hash] The hash of parameters to sanitize
+        # @param [String] The custom redact string
+        # @return [Nil]
+        #
+        def redact!(parts, i, params, redacted_string)
+          return if i == parts.size || !params.key?(parts[i])
```
Comment (mattolson) [inline lib/gruf/instrumentation/request_logging/hook.rb]: Might want to use `i >= parts.size` for more defensive coding.

File: spec/gruf/instrumentation/request_logging/hook_spec.rb [hunk 4]
```diff
@@ -92,6 +92,20 @@ class FakeRequestLogFormatter; end
           expect(subject).to eq expected
         end
       end
+
+      context 'with nested blacklist' do
+        let(:blacklist) { ['data.array', 'hello'] }
+        let(:options) { { blacklist: blacklist } }
+
+        it 'should support nested filtering' do
```
Comment (mattolson) [inline spec/gruf/instrumentation/request_logging/hook_spec.rb]: You should also add a spec for nested hashes.

Comment (jason0415) [inline spec/gruf/instrumentation/request_logging/hook_spec.rb]: It's already added

**Pull request merged**